src/inventory_ui.cpp

```diff
@@ -16,88 +16,396 @@
+                                             const std::string &title )
+{
+    const auto iter = std::find_if( cells.begin(), cells.end(), [ &title ]( const cell_pair & cell ) {
+        return cell.first == title;
+    } );
+    if( iter != cells.end() ) {
+        debugmsg( "Tried to append a duplicate cell \"%s\": ignored.", title.c_str() );
+        return;
+    }
+    cells.emplace_back( title, func );
+}
+
+inventory_column::inventory_column( const inventory_selector_preset &preset ) :
+    preset( preset )
+{
+    cell_widths.resize( preset.get_cells_count(), 0 );
+}
+
+bool inventory_column::activatable() const
+{
+    return std::any_of( entries.begin(), entries.end(), []( const inventory_entry & entry ) {
+        return entry.is_selectable();
+    } );
+}
+
+inventory_entry *inventory_column::find_by_invlet( long invlet )
+{
+    for( auto &entry : entries ) {
         if( entry.is_item() && ( entry.get_invlet() == invlet ) ) {
-            return const_cast<inventory_entry *>( &entry );
+            return &entry;
         }
     }
     return nullptr;
 }
 
-size_t inventory_column::get_width() const
+size_t inventory_column::get_entry_indent( const inventory_entry &entry ) const
 {
-    size_t res = 0;
-    for( const auto &entry : entries ) {
-        res = std::max( res, get_entry_width( entry ) );
+    if( !entry.is_item() ) {
+        return 0;
+    }
+    size_t res = 2;
+    if( OPTIONS["ITEM_SYMBOLS"] ) {
+        res += 2;
+    }
+    if( allows_selecting() && multiselect ) {
+        res += 2;
     }
     return res;
 }
 
-void inventory_column::select( size_t new_index )
+size_t inventory_column::get_entry_cell_width( const inventory_entry &entry, size_t cell_index ) const
+{
+    return preset.get_cell_width( entry, cell_index ) + ( cell_index == 0 ? get_entry_indent( entry ) : min_cell_gap );
+}
+
+void inventory_column::set_width( const size_t width )
+{
+    reset_width();
+    int width_gap = get_width() - width;
+    // Now adjust the width if we must
+    while( width_gap != 0 ) {
+        const int step = width_gap > 0 ? -1 : 1;
+        size_t &cell_width = step > 0
+            ? *std::min_element( cell_widths.begin(), cell_widths.end() )
+            : *std::max_element( cell_widths.begin(), cell_widths.end() );
+        if( cell_width == 0 ) {
+            break; // This is highly unlikely to happen, but just in case
+        }
+        cell_width += step;
+        width_gap += step;
+    }
+}
+
+void inventory_column::set_height( size_t height ) {
+    if( entries_per_page != height ) {
+        if( height == 0 ) {
+            debugmsg( "Unable to assign zero height." );
+            return;
+        }
+        entries_per_page = height;
+        paging_is_valid = false;
+    }
+}
+
+void inventory_column::expand_to_fit( const inventory_entry &entry )
+{
+    for( size_t cell_index = 0; cell_index < cell_widths.size(); ++cell_index ) {
+        size_t &cell_width = cell_widths.at( cell_index );
```

Comment:
Why `vector::at` instead of `operator[]`? Are you expecting an invalid index here? Where is the exception that `at` may throw caught? This also applies to the other usage of `at`.

----

Disclaimer: I'm not against using `at`, it's just a bit inconsistent.

And in this case, the index is valid as per definition in the loop condition, so there is no need to use `at`.